README.md

```diff
@@ -1275,6 +1276,44 @@ Notice that after scrolling to the bottom of the long list of children in the
 Hunting family, then going to another tab and then going to another page, when
 you return to the list of Huntings that the scroll position is maintained.
 
+# Navigator Builder
+Sometimes it is necessary to insert a widget above the `Navigator`, but below `MaterialApp`/`WidgetsApp`: for example, it may be a provider that needs access to the `MaterialApp`/`WidgetsApp` `context` to get the current locale and localization, or to completely replace the `Navigator`, for example to implement authentication with inner navigation.
+
+For these purposes, you need to use the `navigatorBuilder` parameter in the `GoRouter` constructor. This is similar to the `builder` parameter in the `MaterialApp` constructor.
+
+An example of placing some data provider widget:
+```dart
+final _router = GoRouter(
+  // ...routes and errorPageBuilder here...
+
+  // A wrapper around the navigator to provide data.
+  navigatorBuilder: (context, child) => DataProvider(
+    child: child,
+    locale: Localizations.localeOf(context),
+  ),
+);
+```
+
+A more complex example of using `navigatorBuilder` is an implementation of authentication with inner navigation.
+
+```dart
+final _router = GoRouter(
+  // ...routes and errorPageBuilder here...
+
+  // A wrapper around the navigator to implement
+  // login and registration screens.
+  navigatorBuilder: (context, child) => ValueListenableBuilder<bool>(
+    valueListenable: signedIn,
+    child: child,
+    builder: (context, signedInValue, child) => signedInValue
+        ? child
+        : const InnerRouter(child: AuthScreen()),
+  ),
+);
+```
+
+For more details see [`nav_builder.dart`](https://github.com/csells/go_router/blob/master/example/lib/nav_builder.dart).
```

Comment:
I'd like a breakdown of what's happening in this example beyond "read the code," since the code is not at all obvious.